apps/src/blockly/addons/cdoConstantsProvider.js

```diff
@@ -0,0 +1,123 @@
+import GoogleBlockly from 'blockly/core';
+import {blockTypes} from './cdoConstants';
+
+export default class CdoConstantsProvider extends GoogleBlockly.blockRendering
+  .ConstantProvider {
+  /**
+   * Get an object with connection shape and sizing information based on the
+   * type of the connection.
+   *
+   * @param connection The connection to find a shape object for
+   * @returns The shape object for the connection.
+   * @override
+   */
+  shapeFor(connection) {
+    const blockTypeShapeMap = {
+      [blockTypes.SPRITE]: this.TRI_INPUT_OUTPUT,
+      [blockTypes.BEHAVIOR]: this.ROUND_INPUT_OUTPUT,
+      [blockTypes.LOCATION]: this.RECT_INPUT_OUTPUT
+    };
+    // `connection.check_` returns a list of accepted value types for the connection
+    // or null if all types are compatible.
+    // For connections that are customized (sprite, behavior, location), there is
+    // one value type that is accepted so we assign connectorType to the first
+    // element in the list.
+    const connectorType = connection.check_ ? connection.check_[0] : null;
+    switch (connection.type) {
+      case GoogleBlockly.ConnectionType.INPUT_VALUE:
+      case GoogleBlockly.ConnectionType.OUTPUT_VALUE:
+        // PUZZLE_TAB is the default shape for the connector if the value type is not
+        // included in `blockTypeShapeMap`
+        return blockTypeShapeMap[connectorType] || this.PUZZLE_TAB;
+      case GoogleBlockly.ConnectionType.PREVIOUS_STATEMENT:
+      case GoogleBlockly.ConnectionType.NEXT_STATEMENT:
+        return this.NOTCH;
+      default:
+        throw Error('Unknown connection type');
+    }
+  }
+
+  makeTriangularInputConn() {
+    const width = this.TAB_WIDTH;
+    const height = this.TAB_HEIGHT;
+
+    /**
+     * Since input and output connections share the same shape you can
+     * define a function to generate the path for both.
+     */
+    function makeMainPath(up) {
+      return GoogleBlockly.utils.svgPaths.line([
+        GoogleBlockly.utils.svgPaths.point(-width, (-1 * up * height) / 2),
+        GoogleBlockly.utils.svgPaths.point(width, (-1 * up * height) / 2)
+      ]);
+    }
+
+    var pathUp = makeMainPath(1);
+    var pathDown = makeMainPath(-1);
+
+    return {
+      width: width,
+      height: height,
+      pathDown: pathDown,
+      pathUp: pathUp
+    };
+  }
+
+  makeRectangularInputConn() {
+    const width = this.TAB_WIDTH;
+    const height = this.TAB_HEIGHT;
+
+    /**
+     * Since input and output connections share the same shape you can
+     * define a function to generate the path for both.
+     */
+    function makeMainPath(up) {
+      return GoogleBlockly.utils.svgPaths.line([
+        GoogleBlockly.utils.svgPaths.point(-width, 0),
+        GoogleBlockly.utils.svgPaths.point(0, -1 * up * height),
+        GoogleBlockly.utils.svgPaths.point(width, 0)
+      ]);
+    }
+
+    var pathUp = makeMainPath(1);
+    var pathDown = makeMainPath(-1);
```

Comment:
nit: do these need to be `var`s?

Reply:
Mainline uses `var` but I can update to be more consistent with our codebase.